eiffel-syntax-and-usage/activity-linking.md

```diff
@@ -0,0 +1,97 @@
+<!---
+   Copyright 2022 Ericsson AB.
+   For a full list of individual contributors, please see the commit history.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+--->
+
+# Activity Linking
+_Activity linking_ is the act of connecting [activities](../eiffel-syntax-and-usage/glossary.md#activity) within a [pipeline](../eiffel-syntax-and-usage/glossary.md#pipeline) to each other. It includes not just linking pipeline steps to each other, but also linking hierarchically, such as an overall pipeline activity, a set of sub activities for a build or a test, or linking multiple complete pipelines to each other.
+
+Activity links are important both for tracing what activity/activities preceded a certain activity or what activity/activities followed after a certain activity, but also for tracing a full chain of parallel/serial activities within a pipeline. Tracing the activities in pipelines is the base for various kinds of pipeline visualizations and also for metrics and KPI measurements.
+
+## Event Driven vs Orchestrated Pipeline
+A _fully event driven pipeline_ in Eiffel terminology is a pipeline where _all_ activities in it are triggered explicitly by Eiffel events. Those events could either come from earlier pipeline steps, or from other pipelines producing for example artifacts that this pipeline would be configured to trigger on.
+
+A _fully orchestrated pipeline_ is completely controlled by a dedicated pipeline orchestrator, such as Jenkins Pipeline or Argo Workflows, and has no activities triggered by Eiffel events. A fully orchestrated pipeline is probably often initiated by a source change in some SCM system, and that source change is then propagated to the pipeline orchestrator through some non-Eiffel-event channel (e.g. an Github web hook or a Gerrit stream-event)
+
+None of the scenarios above is probably relevant for most of the Eiffel event users, but rather a combination of the two where a pipeline is often _triggered_ by an Eiffel event, but then an orchestrator deals with controlling (at least parts of) the pipeline. Such triggers could for example be SCM events (e.g. [SCC](../eiffel-vocabulary/EiffelSourceChangeCreatedEvent.md)/[SCS](../eiffel-vocabulary/EiffelSourceChangeSubmittedEvent.md)) or artifact events (e.g. [ArtC](../eiffel-vocabulary/EiffelArtifactCreatedEvent.md)/[ArtP](../eiffel-vocabulary/EiffelArtifactPublishedEvent.md)/[CLM](../eiffel-vocabulary/EiffelConfidenceLevelModifiedEvent.md)).
```

Comment:
You mean like a reference section somewhere below the text instead of linking inline? If so, in what way is that more pleasant?

Reply:
Yes, the `[link text][bookmark name]` form of link rather than `[link text](url)`. At least I think that

```
None of the scenarios above is probably relevant for most of the Eiffel event users, but
rather a combination of the two where a pipeline is often _triggered_ by an Eiffel event,
but then an orchestrator deals with controlling (at least parts of) the pipeline. Such triggers
could for example be SCM events
(e.g. [SCC](../eiffel-vocabulary/EiffelSourceChangeCreatedEvent.md)/[SCS](../eiffel-vocabulary/EiffelSourceChangeSubmittedEvent.md))
or artifact events
(e.g. [ArtC](../eiffel-vocabulary/EiffelArtifactCreatedEvent.md)/[ArtP](../eiffel-vocabulary/EiffelArtifactPublishedEvent.md)/[CLM](../eiffel-vocabulary/EiffelConfidenceLevelModifiedEvent.md)).
```

reads better than

```
None of the scenarios above is probably relevant for most of the Eiffel event users, but
rather a combination of the two where a pipeline is often _triggered_ by an Eiffel event,
but then an orchestrator deals with controlling (at least parts of) the pipeline. Such triggers
could for example be SCM events (e.g. [SCC][SCC]/[SCS][SCS]) or artifact events
(e.g. [ArtC][ArtC]/[ArtP][ArtP]/[CLM][CLM]).
```
But either way works, of course.

Reply:
Yeah, I've never used the bookmark notation before but I recently saw it in a different project. I'll try that instead. I agree it is somewhat easier to read, even though I was quite confused as a reader at first about how those magic links where created.